CONTRIBUTING.md

```diff
@@ -51,6 +51,134 @@ If you're *looking* for security bugs, this crate is set up for
+least specific. This means that an `image_id` might go before the `domain`,
+which will go before the `app` context. More specific arguments are more
+differentiating between a given function and other functions, so putting them
+first makes it easier to infer the context/meaning of the function (compared to
+starting with a number of generic context-like types).
+
+#### Error handling
+
+We use `Result` types pervasively throughout the code to signal error cases. We
+prefer to avoid `unwrap()` and `expect()` calls unless there is a clear
+invariant which can be locally validated by the structure of the code. If
+there is such an invariant, we usually add a comment explaining how the
+invariant is upheld. In other cases (especially for error cases which can arise
+from network traffic, which could represent an attacker), we always prefer to
+handle errors and ultimately return an error to the network peer or close the
+connection.
+
+### Expressions
+
+#### Avoid single-use bindings
+
+We generally make full use of the expression-oriented nature of Rust. For
+example, when using iterators we prefer to use `map` and other combinators
+instead of `for`-loops when possible, and will often avoid variable bindings if
+a variable is only used once. Naming variables takes cognitive efforts, and so
+does tracking references to bindings in your mind. One metric we like to
+minimize is the number of mutable bindings in a given scope.
+
+Remember that the overall goal is to make the code easy to understand.
+Combinators can help with this by eliding boilerplate (like replacing a
+`None => None` arm with a `map()` call), but they can also make it harder to
+understand the code. One example is that a combinator chain like
+`.map().map_err()` might be harder to understand than a `match` statement
+(since, in this case, both of the arms have a significant transformation).
+
+#### Use early `return` and `continue` to reduce nesting
+
+The typed nature of Rust can cause some code to end up at deeply indented
+levels, which we call "rightward drift". This makes lines shorter, making the
+code harder to read. To avoid this, try to `return` early for error cases, or
+`continue` early in a loop to skip an iteration.
+
+### Naming
+
+#### Use concise names
+
+We prefer concise names, especially for local variables. Avoid adding a suffix
+for a variable that describes its type (provided that its type is hard to
+confuse with other types -- for example, we do still use `_id` suffixes because
+we usually use numeric IDs for database entities). The precision/conciseness
+trade-off for variable names also depends on the scope of the binding.
+
+#### Avoid `get_` prefixes
+
+Per the
+[API guidelines](https://rust-lang.github.io/api-guidelines/naming.html#getter-names-follow-rust-convention-c-getter),
+`get_()` prefixes are discouraged.
+
+### Imports
+
+We use 3 blocks of imports in our Rust files:
+
+1. `std` imports
```

Comment:
Currently I believe rustls more often uses a `crate`, external, `std`/`core` style, so we can decide which order we prefer. I believe the outside-in style described here is more idiomatic. There is some support in rustfmt and tooling like Rust-Analyzer for splitting up blocks like this.